[1단계 - 로또 구현] 멍구(이유영) 미션 제출합니다.

yuyoungrhee opened the pull request:
### 프롤로그 링크
https://prolog.techcourse.co.kr/studylogs/3875

## 프로젝트 개요
-  페어 프로그래밍 교대 시간: 약 15분
- 공동 목표: 실행은 테스트로만 진행하기
  - 멍구: 단위테스트와 객체 설계에 중점
  - 몽이: 단위테스트 기반 실행 및 기존 코드 참고 최소화
 
## 주요 기술적 의사결정
### 클래스 설계
![IMG_1893](https://github.com/user-attachments/assets/e3c21b65-5a21-47af-8282-f8f80927bd5b)

- Record Class를 활용한 Lotto 객체 구현, Dto 구현
저는 record를 이번 페어프로그래밍에서 처음 써보는데, **불변상태**를 표현할 때 좋다는 페어의 의견이었습니다. 그러나 record는 모든 필드를 public으로 노출한다는 특징이 있는데, 이러한 특성을 가진 record를 domain에 사용해도 되는지 고민이었습니다.
- Wallet(지갑) 객체를 도입하여 Lotto 일급 컬렉션 관리
- 기존 Calculator 도메인 -> Lottoservice
Calculator는 수익률을 계산하는 역할이었는데, 이는 상태가 없어 도메인도 아니고, 유틸이라기에는 비즈니스로직적인 부분이 있었습니다. 결국 서비스로 변경하였습니다. 언제 Service를 도입해야하는지에 대하여 궁금합니다.

### 향후 개선 필요사항
- [x] Wallet 클래스의 책임 분리 검토
  - 로또 생성, 보관, 당첨번호 매칭 책임 분리 고려
- [x] MatchStatistics 클래스명 재검토
- [ ] 예외 발생 시 재입력 처리 로직 고민 (현재 while 문 사용)

## 기술적 고민점
- DependencyConfig 사용
의존성을 일관적으로 관리하기 위해 도입하였고,  둘다 이번 로또 미션에서는 복잡한 의존관계가 없어 필요가 없을 수 있다고 인정하기도 했습니다. 실제로 이런 방법이 미션이 더 복잡해졌을 때 의미가 있을까요?
- 어느정도까지 테스트를 해야하는가에 대한 고민
    - **이미 검증된 객체들을 조합, 또는 반복만 하는 경우 테스트가 필요한가?**
    Wallet은 각 Lotto의 메서드를 사용하여 결과를 생성하는데, 이 경우 Lotto에서 검증했기 때문에 따로 필요가 없다고 판단하여 테스트를 따로 하지 않았습니다.
    - **도메인 로직 상 일어나지 않을 일까지 테스트해야한가?**
    MatchStatistics에서 당첨번호 일치 개수와 보너스여부에 따라 해당 등수를 반환해주는 메서드가 있습니다. 이때 로또 번호 개수는 Lotto에서 검증하고 있기 때문에 matchCount가 -1이 나온다거나 로또번호 개수인 6개를 넘어가는 경우는 실제로 일어나지 않습니다. 이때 MatchStatistics는 온전히 자기 자신의 검증을 책임져야 하기 때문에 이러한 예외까지 처리해야 되는지 고민입니다.

----
# 2번째 PR
### 클래스 설계
![IMG_527C3AA654A6-1](https://github.com/user-attachments/assets/2f0700bd-2542-40fc-94de-5239ad7561df)

Commit:
docs: 기능 요구사항 정리

Commit:
feat: Application 시작점 선언

Commit:
feat: Lotto 객체 선언

Commit:
feat: RandomsWrapper로 랜덤 번호 생성

Commit:
feat: Amount 선언

- 로또 구입 개수를 계산한다.

Commit:
feat: Wallet을 선언

- 고객의 구매한 로또를 관리한다.

Commit:
feat: Controller 선언

Commit:
feat: MatchCountDto를 선언

- 당첨 개수와 보너스 여부를 저장

Commit:
feat: MatchStatistics를 선언

- 당첨 결과에 맞게 변환

Commit:
style: 코딩 컨벤션 적용

Commit:
refactor: 메소드 명 개선 및 코드 구조 정리

- 일부 메소드 명 정정
- 조건문을 메소드로 분리
- Common 패키지 추가

Commit:
feat: Calculator를 선언

- matchCount를 토대로 통계 결과를 반환한다.
- 수익률을 계산한다.

Commit:
feat: Profit Dto를 선언

Commit:
feat: View를 선언

- 입출력 담당 클래스를 선언한다.

Commit:
feat: Parser를 추가한다.

- 입력값을 숫자로 변환한다.
- 입력값을 숫자 List로 반환한다.

Commit:
feat: Wallet의 로또를 전달하기 위해 가공

-toString을 Override하여 수정한다.
-MatchStatics에서 Key에 맞는 String 값을 반환하도록 메소드를 구현한다.

Commit:
feat: Controller.run()을 구현한다.

- DependencyConfig를 통해 의존성 관리

Commit:
refactor: Lotto에서 Match Count를 계산하도록 수정

- Wallet에서 진행하던 일부를 Lotto에게 시켜서 반환하도록 변경했다.
- Constant를 통해 매직 넘버 관리

Commit:
refactor: Calculator를 LottoService로 변경

- 역할이 도메인보다 서비스에 적합하여 변경한다.
- 내부의 메소드명을 정정한다.

Commit:
test: LottoTest에 단위테스트를 추가

- 로또 정렬과 MatchCount()에 대한 테스트를 추가한다.
- InputParser에 String 조건을 상수로 변경한다.

Commit:
feat: 보너스 번호 검증 추가

- 매직넘버, 상수 추가한다.
- 불필요한 테스트 삭제

Commit:
feat: Error 발생 시 다시 질문하기 구현

- MatchStatistics Test를 추가한다.
- Error Message에 Prefix를 추가한다.

Commit:
chore: 스타일 컨밴션 적용

sinb57 commented on src/main/java/lotto/Application.java:
```diff
@@ -0,0 +1,13 @@
+        Controller controller = dependencyConfig.getController();
+        controller.run();
+    }
+}
```
EOF 챙겨주세요~ㅎ

yuyoungrhee replied:
수정했습니다!

sinb57 commented on README.md:
````diff
@@ -1,3 +1,76 @@
+
+
+
+## 실행 결과
+```
+구입금액을 입력해 주세요.
+14000
+14개를 구매했습니다.
+
+[8, 21, 23, 41, 42, 43]
+[3, 5, 11, 16, 32, 38]
+[7, 11, 16, 35, 36, 44]
+[1, 8, 11, 31, 41, 42]
+[13, 14, 16, 38, 42, 45]
````
오 실행 결과도 첨부하셨군요! 👍

sinb57 commented on src/main/java/lotto/common/utill/InputParser.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    private static void validateNumeric(String str) {
+        if (!str.matches(REGULAR_EXPRESSION)) {
```
정규식을 활용하셨군요! 이펙티브자바 (p32. 아이템 6) 참고하시면 좋을 듯 해요~

yuyoungrhee replied:
"불필요한 객체 생성을 피하라"라는 주제와 정규식이 어떤 연관이 있는지 몰랐는데 정규식은 Pattern.compile()를 호출하여 **새로운 Pattern 인스턴스를 만드는 방식**이었습니다. 더군다나 Pattern은 객체 생성 비용이 높은 편에 속하여 성능 면에서 좋지 않다는 것을 배웠습니다.

참고한 글: [[이펙티브 자바] 아이템 6. 불필요한 객체 생성을 피하라](https://velog.io/@lychee/%EC%9D%B4%ED%8E%99%ED%8B%B0%EB%B8%8C-%EC%9E%90%EB%B0%94-%EC%95%84%EC%9D%B4%ED%85%9C-6.-%EB%B6%88%ED%95%84%EC%9A%94%ED%95%9C-%EA%B0%9D%EC%B2%B4-%EC%83%9D%EC%84%B1%EC%9D%84-%ED%94%BC%ED%95%98%EB%9D%BC)

이를 해결하기 위한 방법으로 크게 2가지 방법을 찾았습니다.
#### 1. Pattern 객체를 미리 생성하여 초기 캐싱. 성능을 개선하는 방법
```java
private static final Pattern NUMBER_PATTERN = Pattern.compile("\\d+");

public static void validateNumeric(String str) {
        if (!NUMBER_PATTERN.matcher(str).matches()) {
            throw new IllegalArgumentException("숫자가 아닙니다.");
        }
    }
```

#### 2. Integer.parseInt(str) 방식
```java
private static void validateNumeric(String str) {
        try {
            Integer.parseInt(str);
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException(ERROR_NOT_NUMBER);
        }
    }
```

parseInt() - 숫자로 변환을 시도해보고 안되면 예외라는 뜻으로 더 직관적으로 다가오는 것 같아 2번으로 결정했습니다. 
* 단점은 숫자이지만 Integer 범위를 초과하는 경우에도 같은 NumberFormatException이 발생한다는 것인데, 현재로서는 이런 경우를 고려하지 않기로 했습니다. 혹시 이런 경우까지 고려하여 예외를 더 세세하게 처리해야 할까요?

sinb57 replied:
> 정규식은 Pattern.compile()르 호출하여 새로운 Pattern 인스턴스를 만드는 방식이었습니다.

찾아보거나 라이브러리를 들여다보지 않으면 알기 어렵죠ㅎ그치만 이런 부분들이 성능 저하를 야기하거나 에러를 발생시키기도 합니다. 라이브러리를 사용할 때는 놓친 부분이 없는지 살펴보는게 좋아요~

sinb57 replied:
> 단점은 숫자이지만 Integer 범위를 초과하는 경우에도 같은 NumberFormatException이 발생한다는 것인데, 현재로서는 이런 경우를 고려하지 않기로 했습니다. 혹시 이런 경우까지 고려하여 예외를 더 세세하게 처리해야 할까요?

예외를 상세하게 나눌 수도 있지만 그래야 할까요??
이 상황에서 멍구는 예외를 세세하게 처리해서 어떤 이득을 얻고자 하는지 궁금해요~

yuyoungrhee replied:
예외 케이스와 예외 메시지를 생각하면서 예외를 처리하고 있는데요,현재는 숫자 변환 실패시 `"숫자가 아닙니다."` 라는 메시지를 보여주고 있습니다. 

실제로 숫자인데도 불구하고 범위때문에 변환에 실패하는 경우에 대해서 같은 예외 메시지를 보여주게 된다면 사용자 입장에서 명확하지 않을 수 있을 것 같아요. "유효한 숫자 입력이 아닙니다" 정도로 메시지를 통일하면 해결 가능하지 않을까 합니다. 여기에서 예외를 더 상세하게 나눌 필요는 없는 것 같아요!

sinb57 commented on src/test/java/lotto/utill/RandomsWrapperTest.java:
```diff
@@ -0,0 +1,20 @@
+class RandomsWrapperTest {
+
+    @Test
+    @DisplayName(" ")
```
DisplayName이 비어있네요? 👀

yuyoungrhee replied:
수정했습니다! 참고로 RandomWrapper는 LottoNumberGenerator로 클래스명을 변경하였습니다.

sinb57 commented on src/test/java/lotto/utill/RandomsWrapperTest.java:
```diff
@@ -0,0 +1,20 @@
+    void test() {
+        List<Integer> list = getRandomNumbers();
+        assertThat(list).isNotNull();
+        assertThat(list.size() == 6).isTrue();
```
isTrue 말고 다른 메서드로 바꿀수는 없을까요?

yuyoungrhee replied:
변경했습니다! isEqual()을 쓰면 되는군요

sinb57 commented on src/main/java/lotto/domain/MatchStatistics.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    private static boolean isMatchBonus(int matchNumber, boolean bonus) {
+        return matchNumber == 5 && bonus;
```
매직넘버가 존재하네요ㅎ 의미를 부여하는 다른 방법은 없을까요?

yuyoungrhee replied:
수정했습니다!

다른 상수로 정의하기보다는 보너스 당첨을 위해 일치해야되는 number와 비교하도록 했습니다.

```java
private static boolean isMatchBonus(int matchNumber, boolean bonus) {
        return matchNumber == MATCH_BONUS.number && bonus;
    }
```

sinb57 replied:
MATCH_BONUS 와 연관되도록 하셨군요! 단순히 상수로 분리하는 것보다 좋네요.
다만 의미를 추론해야 하는 부분이 아쉬운데요. 보다 명확히 할 수는 없을까요?

힌트를 드리자면(?) 해당 메서드는 현재 MATCH_BONUS에게만 유의미한 상황이예요.

sinb57 commented on src/main/java/lotto/domain/Amount.java:
```diff
@@ -0,0 +1,43 @@
+
+    private boolean isLessThousand(int money) {
+        return money < LOTTO_PRIZE;
+    }
```
로또 숫자 등의 상수는 변경 가능성이 낮지만 가격은 변동성이 높을텐데요! 가격이 변경되면 상수 뿐만 아니라 메서드명까지도 수정해야 해요~ 한번 고민해보시면 좋을 듯 합니닷

yuyoungrhee replied:
반영했습니다! 상수뿐만 아니라 메서드명도 변경가능성에 주의해야겠네요

sinb57 replied:
`isLessThanLottoPrize` 라는 메서드명으로 수정하셨군요!
이 경우는 LOTTO_PRIZE 상수명 변경에 영향을 받진 않을까 조심스레 생각해보아요~

sinb57 commented on README.md:
```diff
@@ -1,3 +1,76 @@
+   - 랜덤 발급(라이브러리 사용, 6개, 중복 x, 1~45)
+3. 지난주 결과
+    - "지난 주 당첨 번호를 입력해 주세요." 출력
+    - 로또 입력 / 번호 중복 X, 1~45
```
```
지난 주 당첨 번호를 입력해 주세요.
2,2,2,2,2,2
보너스 볼을 입력해주세요.
```
지난 주 당첨 번호를 중복하여 입력할 수 있네요ㅠ 확인 부탁드려요~

yuyoungrhee replied:
로또 내부에서 중복 검증을 추가했습니다! 테스트도 추가했구요

```java
@Test
@DisplayName("로또에 중복된 번호가 있으면 예외가 발생한다.")
void test_validateDuplicate() {
    assertThatThrownBy(() -> new Lotto(new ArrayList<>(List.of(1, 2, 3, 4, 5, 5))))
                .isInstanceOf(IllegalArgumentException.class);
}
```

sinb57 commented on src/main/java/lotto/common/utill/RandomsWrapper.java:
```diff
@@ -0,0 +1,20 @@
+        List<Integer> list = new ArrayList<>();
+
+        while (list.size() < 6) {
+            int randomInt = (int)(Math.random() * 45) + 1;
```
오호 랜덤을 이렇게 구현했군요! 랜덤값이 1-45 사이에 생성된다는 점을 테스트로 커버할 수 없을까요~?

yuyoungrhee replied:
```java
@DisplayName("로또 숫자가 항상 1~45 범위 내에서 생성되는지 테스트 (100번 반복)")
    @RepeatedTest(100)
    void testGeneratedNumbersRange() {
        LottoNumberGenerator numberGenerator = new LottoNumberGenerator();
        List<Integer> lottoNumbers = numberGenerator.generate();

        boolean inRange = lottoNumbers.stream().allMatch(n -> n >= 1 && n <= 45);

        assertThat(inRange).isTrue();
    }
 ```

이런식으로 반복하여 확률적으로 해결하고자 하였는데 더 좋은 방법 있을까요?

sinb57 replied:
현업에서 이뤄질 법한 예시로 힌트 드릴게요! 랜덤이라는 스펙은 기획에 따라, 확률 계산을 달리 해야하는 경우가 발생해요. 확률 분포로 조정되는 경우도 있고요. 랜덤 포인트 지급하는 경우, 신규 가입자에게는 높은 금액이 잘 나오도록 하고, 기존 가입자에게는 랜덤 금액이 나오도록 할 수도 있어요.

~로또라는 특성상 랜덤을 조정하면 불법이지만..~
이처럼 멍구에게 랜덤 확률을 조정하는 기획이 추가된다면, 현재 코드에서 어떻게 적용할 수 있을까요?
OCP를 지키면서 유연한 설계를 해보세요! 그럼 자연스레 테스트도 가능한 구조가 탄생해요.

sinb57 commented on README.md:
```diff
@@ -1,3 +1,76 @@
+3. 지난주 결과
+    - "지난 주 당첨 번호를 입력해 주세요." 출력
+    - 로또 입력 / 번호 중복 X, 1~45
+    - 보너스 번호 입력 / 1개만 / 지난주 번호와 중복 x , 1~45
```
```
보너스 볼을 입력해주세요.
111

당첨 통계
```
숫자 범위를 벗어난 값을 보너스 볼에 입력할 수 있어요~ 확인 부탁드려요

yuyoungrhee replied:
수정했습니다! 예외 처리가 부실했던 것 같습니다..

sinb57 commented on src/main/java/lotto/common/utill/RandomsWrapper.java:
```diff
@@ -0,0 +1,20 @@
+        List<Integer> list = new ArrayList<>();
+
+        while (list.size() < 6) {
+            int randomInt = (int)(Math.random() * 45) + 1;
```
매직넘버를 해소할 수 없을까요?

yuyoungrhee replied:
일부 상수로의 변경이 누락된 것 같습니다. 변경했습니다!

sinb57 commented on src/main/java/lotto/dto/MatchCountDto.java:
```diff
@@ -0,0 +1,4 @@
+package lotto.dto;
+
+public record MatchCountDto(int matchCount, boolean bonus) {
+}
```
Dto가 무엇이며 왜 로직이 없는 클래스를 만들었는지 궁금해요~

yuyoungrhee replied:
Dto는 데이터 전송을 위한 객체이며 흔히 계층 간 데이터 교환을 위해 사용됩니다. 레벨1 순수 자바 미션에서도 MVC 패턴을 적용해서 구현하고 있기 때문에 Controller -> Domain 으로의 데이터 전달을 위해 사용하기도 합니다.

- Controller에서 도메인에게 넘겨줘야하는 값이 여러 개일 때, 이를 하나의 의미있는 묶음으로 전달하고 싶을 때 사용
당첨 정보에는 크게 당첨 번호와 보너스번호가 존재하는데 이를 Dto를 통해 하나로 묶어서 처리하고 있습니다.
```java
//Controller 로직 중
WinningInform winningInform = inputController.getWinningInformation();
LottoStatistics lottoStatistics = LottoStatistics.from(wallet, winningInform);
```

‼️다만 현재 코멘트 위치에 해당하는 MatchCountDto의 경우 계층 간 전달이 아니라 도메인 객체끼리 협력하기 위한 객체였으므로 Dto의 역할에는 해당하지 않는 것 같습니다. 따라서 MatchCount, Profit 등은 도메인 정보를 담아두는 객체이면서 협력할 때 사용되는 구조로써 domain 패키지로 변경하였습니다.

 \+ Dto는 단순 getter를 사용하는 경우가 많기 때문에 record로 구현했습니다. 또한 MatchCount도 도메인에 해당하기는 하지만, 도메인 내부에서 협력을 위한 데이터 묶음, 불변 객체로 볼 수 있어 여전히 record로 구현했습니다.

sinb57 replied:
Dto는 말씀주신대로 데이터 전송을 위한 객체예요!ㅎ
그렇다면 데이터 전송을 위해 왜 Dto를 활용할까요? 도메인 객체를 주고 받으면 안되는 걸까요??

sinb57 commented on src/test/java/lotto/common/utill/InputParserTest.java:
```diff
@@ -0,0 +1,42 @@
+        String str = "";
+        assertThatThrownBy(() -> parseToInt(str))
+            .isInstanceOf(IllegalArgumentException.class);
+    }
```
테스트 설명과 달리 Null이 보이지 않네요.
empty string을 의미한 것이라면 테스트 설명을 의미에 맞춰 수정해주면 좋겠어요~

yuyoungrhee replied:
수정했습니다!

sinb57 commented on src/test/java/lotto/domain/AmountTest.java:
```diff
@@ -0,0 +1,29 @@
+    @DisplayName("정상적인 값을 입력헀을 때 정상적으로 생성된다.")
+    void testAmount() {
+        Amount amount = new Amount(1000);
+        assertThat(amount.getAmount() == 1).isTrue();
```
isTrue 말고 isEqualTo를 사용하면 어떨까요?

yuyoungrhee replied:
수정했습니다!

sinb57 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    private void validateNull(List<Integer> numbers) {
+        if (numbers.isEmpty()) {
```
메서드명과 내부 로직이 상이하네요! 확인 부탁드려요~

yuyoungrhee replied:
수정했습니다!

sinb57 commented on src/main/java/lotto/common/constant/Constant.java:
```diff
@@ -0,0 +1,7 @@
+public class Constant {
+    public static final int LOTTO_MINIMUM = 1;
+    public static final int LOTTO_MAXIMUM = 45;
+    public static final int LOTTO_SIZE = 6;
```
LOTTO_SIZE는 Lotto 객체에서만 사용하는데, Constant 클래스로 분리해놓아야 하나요?
멍구는 상수를 어떻게 관리하는지 궁금해요~

yuyoungrhee replied:
객체 안에서만 사용되는 경우 도메인에서 관리하는 방식과, 로또에 대한 상수들을 한 데 모아두는 방식 중 상황에 따라 선택해서 사용하고 있습니다만 이번에는 모아두는 방식으로 구현해봤습니다.

응집도 측면에서는 사용되는 객체 안에 정의하는 것이 좋다고 생각하는데, 편의 면에서는 모아두는 것이 좋은 것 같기도 합니다. 보통 어느 방식을 선택하나요?

\+ 도메인 안에서 상수를 public으로 열어도 된다면 도메인 안에 정의하는 것이 좋다고 생각합니다. 다른 객체에서 사용되지 않아도 테스트에서 사용하기 때문에 도메인 안에 상수를 private으로 정의하는 점이 불편했습니다.

sinb57 replied:
> 객체 안에서만 사용되는 경우 도메인에서 관리하는 방식과, 로또에 대한 상수들을 한 데 모아두는 방식 중 상황에 따라 선택해서 사용하고 있습니다만 이번에는 모아두는 방식으로 구현해봤습니다.

이번에는 모아두는 방식으로 구현하셨군요!
만약 상수가 많아진다면 어떻게 될까요? 그리고 상수 간의 응집도를 고려해야 할 때는 또 어떻게 해야 할까요?

> 응집도 측면에서는 사용되는 객체 안에 정의하는 것이 좋다고 생각하는데, 편의 면에서는 모아두는 것이 좋은 것 같기도 합니다. 보통 어느 방식을 선택하나요?

도메인이라는 특성에 기반한 응집과, 상수라는 특성에 기반한 응집은 목적성이 달라요. 어떤 기준으로 하는 것이 옳은지 한번 경험해보고 체득하시면 좋겠네요ㅎ

> 도메인 안에서 상수를 public으로 열어도 된다면 도메인 안에 정의하는 것이 좋다고 생각합니다. 다른 객체에서 사용되지 않아도 테스트에서 사용하기 때문에 도메인 안에 상수를 private으로 정의하는 점이 불편했습니다.

도메인이 지닌 상수를 꼭 private으로 정의해야 하나요?

sinb57 commented on src/test/java/lotto/domain/LottoTest.java:
```diff
@@ -0,0 +1,56 @@
+        ArrayList<Integer> list = new ArrayList<>(List.of(6, 5, 4, 3, 2, 1));
+        Lotto lotto = new Lotto(list);
+
+        assertThat(lotto.toString().equals(List.of(6, 5, 4, 3, 2, 1).toString())).isFalse();
```
toString() 처리하지 않고 list의 완전 일치 여부로 검증 할 수는 없을까요?

yuyoungrhee replied:
정렬 확인 테스트를 삭제했습니다.

- Lotto를 기존 record 클래스에서 일반 class로 변경하면서 외부에서 Lotto 내부 번호에 접근할 수 없게 되어 lotto.numbers와 같은 리스트 형태로 로또 번호를 꺼내올 수 없습니다.
- 또한 sort는 단순 Collections 라이브러리를 사용하는데, 라이브러리 자체를 검증할 필요가 없다고 생각해서 삭제했습니다!

sinb57 replied:
Lotto의 생성자를 보면 `sortNumbers(this.numbers)` 로직이 보이는데 정렬에 대한 단위 테스트는 필요가 없을까요?

sinb57 commented on src/main/java/lotto/service/LottoService.java:
```diff
@@ -0,0 +1,44 @@
+import lotto.dto.MatchCountDto;
+import lotto.dto.Profit;
+
+public class LottoService {
```
객체지향으로 해결할 수 없을까요? 현재 미션에서는 Service의 존재가 불필요합니닷

yuyoungrhee replied:
서비스를 삭제하고 LottoStatistics 객체를 추가했습니다. 어제 네오의 수업에서도 Service에 대해서 얘기해보았는데 설계하는 데 고민이 많이 되네요... 

LottoStatistics 객체는 지갑 객체와 당첨정보를 통해서 생성되고, 당첨 개수 계산 / 수익률 계산 등의 통계 관련 기능을 담당합니다.

sinb57 replied:
같은 이치로 Controller 클래스에 대해서도 고민해보시면 좋을 듯 해요ㅎ Controller는 무엇을 책임지는 객체이고, 무엇을 기대하여 Controller라는 이름을 붙이셨나요?

sinb57 commented on src/main/java/lotto/common/config/DependencyConfig.java:
```diff
@@ -0,0 +1,21 @@
+import lotto.view.InputView;
+import lotto.view.OutputView;
+
+public class DependencyConfig {
+    private Controller controller;
```
> 의존성을 일관적으로 관리하기 위해 도입하였고, 둘다 이번 로또 미션에서는 복잡한 의존관계가 없어 필요가 없을 수 있다고 인정하기도 했습니다. 실제로 이런 방법이 미션이 더 복잡해졌을 때 의미가 있을까요?

DependencyConfig를 통해 의존성을 관리했을 때, 멍구가 생각하는 장단점이 무엇인지 궁금해요~!

yuyoungrhee replied:
우선 현재 미션 기준으로는 단점으로 느껴지기도 하는데, 그 이유는 오히려 코드가 길어져서 생산성이 좋지 않은 것 같습니다. 각 객체에서 new로 생성하기만 하면 간단한데, Config라는 파일을 같이 확인해야하는 점이 흐름 파악 면에서 방해가 될 수 도 있을 것 같습니다.

장점으로는 의존성이 복잡한 프로젝트의 경우 오히려 한 곳에 모아두는 것이 의존 관계를 파악하기가 쉬울 것 같습니다.. 다만 의존성이 복잡하다는 것 자체가 좋지 않은 코드라는 신호를 나타내는 것 같기도 하네요.. 

단순히 객체가 많아서 의존관계 파악이 필요한 경우에 DependeyConfig를 사용하면 어떨까요? 미션하면서 그런 상황을 만나게 된다면 다시 고민해보고 싶습니다..

sinb57 replied:
생각을 공유해주셔서 감사해요ㅎ 말씀주셨듯 현재 미션 기준으로는 DependencyConfig가 필요없어 보여요.
의존성 관리가 필요한 수준의 복잡도가 아니거든요. 그렇지만 좋은 시도였어요ㅎ

그치만 덕분에 의존성 주입이 열린 구조로 설계됐는데요. 의존성 주입이 열리면 어떤 장점이 있을까요?

sinb57 commented on src/main/java/lotto/domain/Wallet.java:
```diff
@@ -0,0 +1,42 @@
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
```
StringBuilder를 사용했군요! 👍

sinb57 commented on src/main/java/lotto/domain/Wallet.java:
```diff
@@ -0,0 +1,42 @@
+public class Wallet {
+    private final List<Lotto> lottos = new ArrayList<>();
+
+    public Wallet(Amount amount) {
+        int lottoAmount = amount.getAmount();
+        for (int i = 0; i < lottoAmount; i++) {
+            lottos.add(new Lotto(getRandomNumbers()));
+        }
+    }
```
> Wallet은 각 Lotto의 메서드를 사용하여 결과를 생성하는데, 이 경우 Lotto에서 검증했기 때문에 따로 필요가 없다고 판단하여 테스트를 따로 하지 않았습니다.

좋은 생각이에요! Lotto가 지닌 책임은 무엇이고 Wallet이 지닌 책임은 무엇인지 고려하면서, 각 책임에 대한 테스트를 작성하는게 좋아요. 그런 면에서, wallet이 옳은 개수만큼 로또를 들고 있는지 테스트를 추가하면 좋겠어요ㅎ

yuyoungrhee replied:
로또를 구매하는 책임을 LottoShop으로 분리하면서 개수대로 구매되는지의 테스트도 LottoShopTest에 추가하였습니다.

sinb57 commented on src/test/java/lotto/domain/MatchStatisticsTest.java:
```diff
@@ -0,0 +1,18 @@
+
+class MatchStatisticsTest {
+
+    @Test
+    @DisplayName("getMatchStatistics 정상 작동 테스트")
+    void testGetMatchStatistics() {
+        assertThat(getMatchStatistics(3, false)).isEqualTo(MATCH_THREE);
+        assertThat(getMatchStatistics(5, false)).isEqualTo(MATCH_FIVE);
+        assertThat(getMatchStatistics(5, true)).isEqualTo(MATCH_BONUS);
+    }
```
ParameterizedTest를 찾아보시면 딱 좋을 듯 해요~!

yuyoungrhee replied:
적용 완료했습니다!👍

모든 케이스에 대한 테스트를 해도 테스트가 깔끔해져 좋은 것 같습니다. 3가지 경우 말고도 모든 경우에 대해서 정상 작동되는지 테스트했습니다.

sinb57 commented on src/main/java/lotto/domain/MatchStatistics.java:
```diff
@@ -0,0 +1,49 @@
+        return String.format(BASIC_OUTPUT, number, money);
+    }
+
+    public static MatchStatistics getMatchStatistics(int matchNumber, boolean bonus) {
```
> MatchStatistics에서 당첨번호 일치 개수와 보너스여부에 따라 해당 등수를 반환해주는 메서드가 있습니다. 이때 로또 번호 개수는 Lotto에서 검증하고 있기 때문에 matchCount가 -1이 나온다거나 로또번호 개수인 6개를 넘어가는 경우는 실제로 일어나지 않습니다. 이때 MatchStatistics는 온전히 자기 자신의 검증을 책임져야 하기 때문에 이러한 예외까지 처리해야 되는지 고민입니다.

모호한 상황에서는 접근을 달리해봐도 좋아요~ int matchNumber, boolean bonus를 외부에서 계산하여 넘겨받는게 아니라, 내부에서 계산하도록 처리하면 되지 않을까요? 그렇다면 예기치못한 인자 값이 들어온다는 케이스를 근본적으로 해소할 수 있어 보이네요

yuyoungrhee replied:
가장 고민되는 코멘트였습니다..!!

내부에서 계산하도록 처리하면 MatchStatistics(현재 MatchRank)가 매칭까지 처리하게 되어 enum 클래스가 너무 커질 수 있겠다는 우려가 있었습니다. 따라서 우선 현재 설계를 유지하려고 합니다.

 다만 실제로 예외가 발생하는 인자가 들어오면 예외를 발생시키도록 validate를 추가했습니다.

```java
public static MatchRank getMatchRank(int matchNumber, boolean bonus) {
        validateMatchNumber(matchNumber); //추가

        if (isMatchBonus(matchNumber, bonus)) {
            return MATCH_BONUS;
        }

        for (MatchRank statistics : values()) {
            if (statistics.number == matchNumber) {
                return statistics;
            }
        }

        return NO_MATCH;
    }

```

```java
private static void validateMatchNumber(int matchNumber) {
        if(matchNumber < MIN_MATCH_NUMBER || matchNumber > MAX_MATCH_NUMBER) {
            throw new IllegalArgumentException(ERROR_MATCH_NUMBER_NOT_VALID);
        }
    }
```

sinb57 replied:
MatchResult를 인자로 넘겨받으면 검증에 대한 책임을 MatchRank가 지니지 않아도 되지 않을까요??
enum 클래스가 커진다면 책임 분리에 문제가 있는 것이기에 도메인 설계를 다시 짚어보는 것도 좋습니다!

sinb57 reviewed (changes requested):
안녕하세요 멍구~ 로또 미션의 리뷰어를 맡은 이프 입니다! 잘 부탁드려요~ㅎ

좋은 코드 작성해주셔서 재밌게 봤습니다ㅎ 실행해보니 요구사항과 다른 결과가 나온 부분이 있어서 코멘트 남겨두었어요! 확인하시고 반영 부탁드려요~

남겨주신 고민점에 대해 각각의 코드 위치에 코멘트 남겼는데요, 이 부분도 확인해보시고 필요하다면 추가 질문 주세요~

yuyoungrhee committed:
style: Application 클래스 EOF 추가

yuyoungrhee committed:
refactor(InputParser): 숫자 변환 구현을 정규식에서 Integer.parseInt()로 수정

- 정규식은 내부적으로 Pattern이라는 인스턴스 객체를 사용하는, 불필요한 객체를 생성하는 형태이기 때문에 수정함

yuyoungrhee committed:
test(InputParser): 리스트 변환 테스트 추가, 테스트 일부 수정

-parseToInt()뿐만 아니라 pareToList() 메서드의 테스트도 추가
- testNullError라는 테스트 이름에 맞게 ""이 아닌 null 테스트로 수정

yuyoungrhee committed:
refactor(Wallet): 로또들의 일치 결과를 반환하는 메서드 네이밍 변경

조금 더 직관적인 네이밍으로 변경 (matchCount -> getMatchResults)

yuyoungrhee committed:
refactor(Constant): 상수 이름 네이밍 변경

- 로또의 최소범위, 최대범위라는 뜻의 네이밍을 더 직관적으로 하고자 RANGE라는 단어를 추가

yuyoungrhee committed:
refactor(Lotto): 검증 로직 추가, 구조 변경 등

- Lotto는 캡슐화된 도메인으로 취급하는 것이 더 좋다고 판단하여 record -> class로 변경함.
- 메서드명 변경, matchCount -> countMatchingNumbers : 얼마나매칭되는지 "센다" 기 때문에 count 동사로 시작하는 메서드명으로 변경 필요성을 느낌
- record -> class 변경으로 인해 matchLotto의 numbers를 get해오지 못하기 때문에 캡슐화 하여 비교할 수 있도록 countMatchingNumbers 메서드 로직 수정
- 로또 내에 중복된 번호가 있는지 검증 추가, 테스트 추가
- 에러 메시지 추가 및 변수 네이밍 수정

yuyoungrhee committed:
test(Lotto): 정렬 테스트 삭제

- Lotto를 record 객체에서 일반 class로 바꾸면서 Lotto 안에서 보관하고 있는 번호들의 getter가 삭제됨, 이에 따라 테스트 불가 + Collections 라이브러리의 정렬 기능을 굳이 테스트할 필요 없음

yuyoungrhee committed:
refactor(Amount): 메서드 명 변경

- 로또 가격은 변경 가능성이 있음
로또 하나의 가격보다 돈이 적은지를 반환하는 메서드 명을 Thousand라는 가격을 직접적으로 나타내는 이름에서 LottoPrize로 변경.

yuyoungrhee committed:
refactor: 랜덤 숫자 생성 클래스명 변경 및 리팩토링

- LottoNumberGenerator로 클래스 이름 변경: RandomWrapper에서 일반적인 네이밍인 Generator로 네이밍을 변경함

- 매직 넘버 수정
- 테스트: isEqaulTo를 사용하도록 수정, 범위 테스트 추가
- 상수명 변경 LOTTO_SIZE -> LOTTO_NUMBER_COUNT 해당 상수가 번호의 "개수"를 의미하는 것을 명확히 하도록 변경

yuyoungrhee committed:
refactor: Amount -> Money 클래스명 변경

Money에서 살수있는 로또 개수를 반환해주는 것이 더 자연스럽다고 생각해서 클래스명 변경

- 메서드명 변경 validateZero -> validateEnoughMoney : money가 0일 때 말고도 최소가격을 넘지 못하면 예외를 발생시키는 메서드였으므로 의미에 맞게 메서드명을 변경
- isEqualTo를 사용하여 테스트하도록 변경

yuyoungrhee committed:
refactor: LottoNumberGenerator를 도메인으로 변경

단순히 랜덤한 번호를 생성하는 것이 아니라 로또 규칙(6개, 1~45 범위 등)에 따라서 생성하므로 도메인에 가깝다고 생각함
- 패키지 util -> domain 이동
- static 메서드 -> 일반 메서드

yuyoungrhee committed:
refactor(Wallet): 로또를 개수에 맞게 구매하는 책임 분리

Wallet은 단순히 로또 리스트를 받아서 보관하고, 로또들에 대한 결과를 반환해주는 역할만 수행
- Controller에서 LottoShop에 money를 제공해서 Wallet을 반환받는 형태로 수정

yuyoungrhee committed:
refactor(LottoShop): 메서드 반환값 수정

- 테스트 용이성 및 메서드명과 메서드가 하는 일의 일치를 위하여 변경

yuyoungrhee committed:
refactor(LottoNumberGenerator): 필요없는 주석 제거

yuyoungrhee committed:
refactor: MatchStatistics -> Rank로 enum 클래스명 변경

yuyoungrhee committed:
refactor(MatchRank): 매직넘버 수정

yuyoungrhee committed:
refactor: 서비스 삭제 및 객체 추가

LottoService를 삭제하고 통계를 관리하고 관련 기능을 제공하는 LottoStatistics 객체 추가

yuyoungrhee committed:
refactor(MatchCountDto): 변수명 변경

- boolean 속성을 더 잘 확인할 수 있도록 변경함

yuyoungrhee committed:
refactor: MatchStatistics 변수명 변경 -> MatchRank

yuyoungrhee committed:
refactor(Profit): 정적 팩토리 메서드 정의

- 이미 계산된 rate와 isProfit 정보를 받아오는 것이 아니라, 당첨금과 소비금을 인자로 받아 스스로 Profit을 생성할 수 있도록 함

yuyoungrhee committed:
refactor(Profit): 정적 팩토리 메서드 정의

- 이미 계산된 rate와 isProfit 정보를 받아오는 것이 아니라, 당첨금과 소비금을 인자로 받아 스스로 Profit을 생성할 수 있도록 함

yuyoungrhee committed:
refactor(view): 설계 변경에 따른 view 수정

yuyoungrhee committed:
refactor(controller): Input에 관련된 Controller를 분리

- Controller와, input에 대한 컨트롤러를 따로 분리함 -> InputController 구현

yuyoungrhee committed:
test(Profit): 테스트 추가, 경계값 테스트 추가

yuyoungrhee committed:
test(MatchRankTest): @ParameterizedTest를 이용한 테스트로 변경

yuyoungrhee committed:
refactor: dto 네이밍 변경 및 패키지 변경

Dto의 역할보다는 단순히 도메인 값 객체데 가까우므로 네이밍을 변경함
- dto패키지에 있지만 domain 역할을 하고있는 클래스를 도메인 패키지로 옮김

yuyoungrhee committed:
refactor(MatchRank): 예외처리 추가

yuyoungrhee committed:
refactor(MatchRank): 메서드 분리 및 에러메시지 상수화

sinb57 commented on src/main/java/lotto/domain/Money.java:
```diff
@@ -0,0 +1,42 @@
+import static lotto.common.constant.Constant.LOTTO_PRIZE;
+import static lotto.common.exception.ErrorMessage.*;
+
+public class Money {
+    private final int money;
```
원시값을 포장해주셨군요!
이와 같이 원시값을 포장할 수 있는 여지가 아직 도메인 상에 존재하는데요, 찾아서 적용해보시겠어요?

sinb57 replied:
Money와 같이 원시값을 포장할 수 있는 부분이 다른 곳에 또 없을까요?

sinb57 commented on src/main/java/lotto/common/exception/ErrorMessage.java:
```diff
@@ -0,0 +1,15 @@
+    private static final String PREFIX = "[ERROR] ";
+    public static final String ERROR_NUMBER_RANGE = PREFIX + "로또 번호는 1에서 45 사이여야 합니다.";
+    public static final String ERROR_LOTTO_SIZE = PREFIX + "로또 번호는 6개여야 합니다";
+    public static final String ERROR_MONEY_ZERO = PREFIX + "로또를 구매하려면 천 원이상 필요합니다.";
```
ERROR_MONEY_ZERO 라는 상수명을 봤을 때 0과 관련된 메시지라고 추측했는데, 에러 메시지는 이와 다르네요ㅎ
의미를 왜곡할 수도 있는 상황이라면, 추상화하지 않는 것(상수 분리)도 하나의 방법이에요~

sinb57 commented on src/main/java/lotto/controller/InputController.java:
```diff
@@ -0,0 +1,98 @@
+        }
+    }
+
+    private void validateBonusRange(int bonus) {
+        if(bonus < LOTTO_RANGE_MINIMUM || bonus > LOTTO_RANGE_MAXIMUM) {
+            throw new IllegalArgumentException(ERROR_BONUS_OUT_OF_RANGE);
+        }
+    }
```
보너스 번호의 숫자 범위 검증 로직을 추가하셨군요!
Lotto 안의 검증 로직과 함께 중복을 제거할 수는 없을까요?

sinb57 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,94 @@
+        validateDuplicate(numbers);
+    }
+
+    private static void validateNull(List<Integer> numbers) {
+        if (numbers == null) {
+            throw new IllegalArgumentException(ERROR_LOTTO_SIZE);
+        }
+    }
```
null을 체크하시는군요! Integer에 대해서는 null 체크를 하지 않아도 될까요? null을 체크하는 이유가 궁금해요~

sinb57 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    private static void validateDuplicate(List<Integer> numbers) {
+        Set<Integer> set = new HashSet<>();
```
자료구조에 의존하는 변수명은 좋지 않아요~

sinb57 commented on src/main/java/lotto/view/InputView.java:
```diff
@@ -0,0 +1,21 @@
+
+import java.util.Scanner;
+
+public class InputView implements AutoCloseable {
```
AutoCloseable 인터페이스의 구현체로 만든 이유가 있을까요?

sinb57 commented on src/main/java/lotto/view/InputView.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.Scanner;
+
+public class InputView implements AutoCloseable {
+    private Scanner scanner;
+
+    public InputView() {
+        this.scanner = new Scanner(System.in);
+    }
```
생성자에서 Scanner를 생성해야 하나요?

sinb57 commented on src/test/java/lotto/domain/LottoStatisticsTest.java:
```diff
@@ -0,0 +1,60 @@
+        assertThat(profit.isProfit()).isFalse();
+    }
+
+}
```
EOF가 빠진 곳이 여럿 보이네요ㅎ 👀

sinb57 commented on src/test/java/lotto/domain/LottoTest.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    @Test
+    @DisplayName("MatchCount()에서 로또 당첨 개수와 보너스 여부를 반환한다.")
+    void test_MatchCount() {
+        Lotto lotto = new Lotto(new ArrayList<>(List.of(1, 2, 3, 4, 5, 6)));
+        Lotto matchLotto = new Lotto(new ArrayList<>(List.of(1, 2, 3, 4, 5, 8)));
+        int bonus = 6;
+        MatchResult dto = lotto.countMatchingNumbers(matchLotto, bonus);
+
+        assertThat(dto.matchCount()).isEqualTo(5);
+        assertThat(dto.isBonusMatched()).isTrue();
+    }
```
테스트 설명과 테스트 구현이 상이하네요! 확인 부탁드려요~

sinb57 commented on src/test/java/lotto/domain/LottoTest.java:
```diff
@@ -0,0 +1,53 @@
+    @Test
+    @DisplayName("로또 생성 검증")
+    void testLotto() {
+        assertThat(new Lotto(new ArrayList<>(List.of(1, 2, 3, 4, 5, 6)))).isNotNull();
```
ArrayList로 재생성하는 이유가 궁금해요~

sinb57 reviewed (approved):
안녕하세요 멍구~
답변 주신 부분에 대해 코멘트 남겨두었는데 확인 부탁드려요
리팩터링에 대한 리뷰로 넘어가기 위해 1단계는 머지하겠습니다ㅎ
